### PR TITLE
Develop remove flux numerical derivatives

### PR DESCRIPTION
**Pull request by seantrim**
Make sure all the relevant boxes are checked (and only check the box if you actually completed the step):

- [ ] Closes #xxx (identify the issue associated with this PR)
- [x] Code passes standard test cases (results are either bit-for-bit identical, or differences are explained in the PR comment)
- [ ] New tests added (describe which tests were performed to test the changes)
- [ ] Science test figures (add figures to PR comment and describe the tests)
- [ ] Checked that the new code conforms to the [SUMMA coding conventions](https://github.com/NCAR/summa/blob/master/docs/howto/summa_coding_conventions.md)
- [x] Describe the change in the release notes (use either `./summa/docs/whats-new.md` or `./summa/docs/minor-changes.md` depending on what changed)

This pull request pertains to the removal of numerical derivatives from flux calculations as requested by @martynpclark. The numerical derivatives were previously used to test the analytical formulas for the derivatives. Now that analytical derivatives are available for all cases in the flux files, the numerical derivative code sections could be removed, making the code more concise and more efficient (due to a reduction in branching). If the user happens to request numerical derivatives for fluxes in the model decisions, an error message is returned explaining that numerical derivatives are no longer available. Model output is bit identical to the current develop branch (verified using the full suite of 35 tests). Hopefully this does not interfere with @ashleymedin's recent pull request too much.

### CODE REVIEW

File: build/source/engine/soilLiqFlx.f90 [hunk 1]
```diff
@@ -194,6 +192,7 @@ subroutine soilLiqFlx(&
  ! local variables: general
  character(LEN=256)                  :: cmessage                     ! error message of downwind routine
  integer(i4b)                        :: ibeg,iend                    ! start and end indices of the soil layers in concatanated snow-soil vector
+ logical(lgt),parameter              :: deriv_desired=.true.         ! flag indicating if derivatives are desired
```
Comment (wknoben) [inline build/source/engine/soilLiqFlx.f90:194]: I believe this is the first use of default function argument values in the code. Is that correct? 

If so, with default arguments we trade readability (user sees less text while reading the code) for clarity (the fact that an argument to the function is available at all becomes hidden). We don't currently have easily accessible function-level documentation that a user can check to learn more about individual functions. In this light, is this trade-off worth it?

File: docs/whats-new.md [hunk 2]
```diff
@@ -25,3 +25,4 @@ This page provides simple, high-level documentation about what has changed in ea
 - Flux files have been refactored (code shortened by nearly 900 lines)
 - Fortran arithmetic in flux files has been updated to increase accuracy and reduce computational cost
 - Script added (summa/build/compile_SUMMA.sh) to simplify the compilation process
+- Argument list for soilLiqFlx subroutine has been reduced
```
Comment (wknoben) [inline docs/whats-new.md]: I would like this to be more specific:
```
Argument <x> of `soilLiqFlx()` has been made optional and assigned default value `True`; argument removed from function calls in function <y>, <z>
```

File: build/source/engine/soilLiqFlx.f90 [hunk 3]
```diff
@@ -297,27 +283,13 @@ subroutine soilLiqFlx(&
...
-  desireAnal = .false.
- end if
-
- ! check the need to compute numerical derivatives
- if (deriv_desired .and. ixDerivMethod==numerical) then
-  nFlux=3  ! compute the derivatives using one-sided finite differences
- else
-  nFlux=0  ! compute analytical derivatives
- end if
+ desireAnal = .true. ! set flag for computing analytical derivatives
```
Comment (wknoben) [inline build/source/engine/soilLiqFlx.f90:292]: Seems weird to set a flag to always be true. Can this flag and, if needed, any reliance on this flag later in the routine be removed?

File: docs/whats-new.md [hunk 4]
```diff
@@ -22,7 +22,8 @@ This page provides simple, high-level documentation about what has changed in ea
 - Fixes a bug that incorrectly writes scalarRainPlusMelt to output in cases where snow layers do not exist
 - Changed part "(a,1x,i0)" to "(a,1x,i0,a,f5.3,a,f5.3)" in check_icond.f90 line 277 to print out error correctly.
 - Fixes a bug where restart files are not read correctly in cases where the parallelization argument `-g` for setups that have >1 HRU per GRU
-- Flux files have been refactored (code shortened by nearly 1100 lines)
+- Flux files have been refactored (code shortened by nearly 1500 lines)
 - Fortran arithmetic in flux files has been updated to increase accuracy and reduce computational cost
 - Script added (summa/build/compile_SUMMA.sh) to simplify the compilation process
-- Argument list for soilLiqFlx subroutine has been reduced
+- Argument lists for soilLiqFlx, aeroResist, aStability, bulkRichardson, and turbFluxes subroutines have been reduced
+- Numerical derivatives that were no longer used were removed from flux calculation routines
```
Comment (wknoben) [inline docs/whats-new.md:30]: Can this be updated to specify from which routines the numerical derivatives were removed?

File: docs/whats-new.md [hunk 5]
```diff
@@ -22,7 +22,8 @@ This page provides simple, high-level documentation about what has changed in ea
 - Fixes a bug that incorrectly writes scalarRainPlusMelt to output in cases where snow layers do not exist
 - Changed part "(a,1x,i0)" to "(a,1x,i0,a,f5.3,a,f5.3)" in check_icond.f90 line 277 to print out error correctly.
 - Fixes a bug where restart files are not read correctly in cases where the parallelization argument `-g` for setups that have >1 HRU per GRU
-- Flux files have been refactored (code shortened by nearly 1100 lines)
+- Flux files have been refactored (code shortened by nearly 1500 lines)
 - Fortran arithmetic in flux files has been updated to increase accuracy and reduce computational cost
 - Script added (summa/build/compile_SUMMA.sh) to simplify the compilation process
-- Argument list for soilLiqFlx subroutine has been reduced
+- Argument lists for soilLiqFlx, aeroResist, aStability, bulkRichardson, and turbFluxes subroutines have been reduced
```
Comment (wknoben) [inline docs/whats-new.md:29]: Can all of these to be a bit more detailed, as described in an earlier review comment?

File: build/source/engine/vegNrgFlux.f90 [hunk 6]
```diff
@@ -252,18 +251,10 @@ subroutine vegNrgFlux(&
...
- real(rkind)                        :: canopyTemp                               ! value of canopy temperature used in flux calculations (may be perturbed)
- real(rkind)                        :: canairTemp                               ! value of canopy air temperature used in flux calculations (may be perturbed)
- real(rkind)                        :: try0,try1                                ! trial values to evaluate specific derivatives (testing only)
+ ! local (temperature values used in flux calculations)
+ real(rkind)                        :: groundTemp                               ! value of ground temperature used in flux calculations
+ real(rkind)                        :: canopyTemp                               ! value of canopy temperature used in flux calculations
+ real(rkind)                        :: canairTemp                               ! value of canopy air temperature used in flux calculations
  ! local (saturation vapor pressure of veg)
  real(rkind)                        :: TV_celcius                               ! vegetaion temperature (C)
```
Comment (wknoben) [inline build/source/engine/vegNrgFlux.f90:259]: Can we fix this "vegetaion" typo while we're at it?

File: build/source/engine/vegNrgFlux.f90 [hunk 7]
```diff
@@ -538,6 +492,11 @@ subroutine vegNrgFlux(&
  ! ---------------------------------------------------------------------------------------
  err=0; message="vegNrgFlux/" ! initialize error control
 
+ ! Validate method for calculation of derivatives
+ if (ix_fDerivMeth /= analytical) then
+  err=20; message=trim(message)//'only analytical derivatives are supported -- numerical derivatives are no longer used'; return
+ end if
```
Comment (wknoben) [inline build/source/engine/vegNrgFlux.f90:498]: So far I think I've seen this error as a single line, 4 lines and now 3 lines. Is there a reason to choose any particular one? Might be neater to make this consistent across all files

**Comment by seantrim** [comment]
Thanks for you comments Wouter! Based on today's merger meeting, I will close this PR for now, as it depends on the previous flux refactoring PR that will be reverted. Additionally, Ashley's changes may already cover the numerical derivative removal from the flux routines. I will revisit this PR after the big merger is complete. If needed, I will resubmit the PR to the upcoming v4 branch and resolve your comments above if they still apply. I will also make sure that the numerical derivatives removed here are covered by Ashley's changes.

**Pull request closed without merging**